File: labs/packages/csharp-sdk/API/DocumentClient.cs
using SchlepEngine.Types;

namespace SchlepEngine.API;

/// <summary>
/// Client for document extraction operations.
/// </summary>
/// <remarks>
/// Provides methods for extracting text, tables, images, and metadata from documents
/// using OCR and advanced extraction techniques.
/// </remarks>
/// <example>
/// <code>
/// var client = new SchlepClient("your-api-key");
/// var fileData = await File.ReadAllBytesAsync("document.pdf");
/// var result = await client.Document.ExtractTextAsync(fileData, "document.pdf");
/// Console.WriteLine($"Extracted text: {result.Text}");
/// </code>
/// </example>
public class DocumentClient
{
    private readonly SchlepClient _client;

    /// <summary>
    /// Initializes a new instance of the DocumentClient class.
    /// </summary>
    /// <param name="client">The parent SchlepClient instance.</param>
    internal DocumentClient(SchlepClient client)
    {
        _client = client;
    }

    /// <summary>
    /// Extract text from a document.
    /// </summary>
    /// <param name="fileData">Document file data.</param>
    /// <param name="filename">Document filename.</param>
    /// <param name="extractTables">Whether to extract tables (default: true).</param>
    /// <param name="extractImages">Whether to extract images (default: false).</param>
    /// <param name="cancellationToken">Cancellation token.</param>
    /// <returns>Extraction results with text and metadata.</returns>
    /// <exception cref="ApiException">Thrown if the API returns an error.</exception>
    public async Task<ExtractionResponse> ExtractTextAsync(
        byte[] fileData,
        string filename,
        bool extractTables = true,
        bool extractImages = false,
        CancellationToken cancellationToken = default)
    {
        var form = new MultipartFormDataContent();
        form.Add(new ByteArrayContent(fileData), "file", filename);
        form.Add(new StringContent(extractTables.ToString().ToLower()), "extract_tables");
        form.Add(new StringContent(extractImages.ToString().ToLower()), "extract_images");

        return await _client.PostMultipartAsync<ExtractionResponse>("/extract/text", form, cancellationToken);
    }

    /// <summary>
    /// Extract tables from a document.
    /// </summary>
    /// <param name="fileData">Document file data.</param>
    /// <param name="filename">Document filename.</param>
    /// <param name="cancellationToken">Cancellation token.</param>
    /// <returns>Extracted tables with structured data.</returns>
    /// <exception cref="ApiException">Thrown if the API returns an error.</exception>
    public async Task<TableExtractionResponse> ExtractTablesAsync(
        byte[] fileData,
        string filename,
        CancellationToken cancellationToken = default)
    {
        var form = new MultipartFormDataContent();
        form.Add(new ByteArrayContent(fileData), "file", filename);

        return await _client.PostMultipartAsync<TableExtractionResponse>("/extract/tables", form, cancellationToken);
    }

    /// <summary>
    /// Extract images from a document.
    /// </summary>
    /// <param name="fileData">Document file data.</param>
    /// <param name="filename">Document filename.</param>
    /// <param name="cancellationToken">Cancellation token.</param>
    /// <returns>Extracted images with metadata.</returns>
    /// <exception cref="ApiException">Thrown if the API returns an error.</exception>
    public async Task<ImageExtractionResponse> ExtractImagesAsync(
        byte[] fileData,
        string filename,
        CancellationToken cancellationToken = default)
    {
        var form = new MultipartFormDataContent();
        form.Add(new ByteArrayContent(fileData), "file", filename);

        return await _client.PostMultipartAsync<ImageExtractionResponse>("/extract/images", form, cancellationToken);
    }

    /// <summary>
    /// Perform OCR on an image or document.
    /// </summary>
    /// <param name="fileData">Image or document file data.</param>
    /// <param name="filename">File name.</param>
    /// <param name="language">OCR language (default: "en").</param>
    /// <param name="cancellationToken">Cancellation token.</param>
    /// <returns>OCR results with recognized text.</returns>
    /// <exception cref="ApiException">Thrown if the API returns an error.</exception>
    public async Task<OCRResponse> PerformOCRAsync(
        byte[] fileData,
        string filename,
        string language = "en",
        CancellationToken cancellationToken = default)
    {
        var form = new MultipartFormDataContent();
        form.Add(new ByteArrayContent(fileData), "file", filename);
        form.Add(new StringContent(language), "language");

        return await _client.PostMultipartAsync<OCRResponse>("/extract/ocr", form, cancellationToken);
    }
}

File: labs/packages/java-sdk/src/main/java/io/schlepengine/api/DocumentClient.java
package io.schlepengine.api;

import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import io.schlepengine.exceptions.ApiException;
import okhttp3.OkHttpClient;

import java.io.File;
import java.io.IOException;
import java.util.HashMap;
import java.util.Map;

/**
 * Document Extraction API client.
 *
 * <p>Provides methods for extracting text, tables, and metadata from documents.</p>
 *
 * <h3>Supported Document Types</h3>
 * <ul>
 *   <li>PDF documents</li>
 *   <li>Word documents (.doc, .docx)</li>
 *   <li>Excel spreadsheets (.xls, .xlsx)</li>
 *   <li>Images with OCR (PNG, JPEG, TIFF)</li>
 * </ul>
 *
 * <h3>Example Usage</h3>
 * <pre>{@code
 * DocumentClient docClient = client.document();
 *
 * // Extract text from document
 * File pdfFile = new File("document.pdf");
 * Map<String, Object> result = docClient.extractText(pdfFile, true, false);
 *
 * // Extract tables
 * Map<String, Object> tables = docClient.extractTables(pdfFile);
 * }</pre>
 *
 * @author Schlep-engine Team
 * @version 1.0.0
 * @since 1.0.0
 */
public class DocumentClient extends BaseClient {

    /**
     * Create a new document extraction client.
     *
     * @param httpClient HTTP client instance
     * @param objectMapper JSON mapper instance
     * @param baseUrl API base URL
     * @param apiKey API authentication key
     */
    public DocumentClient(
        OkHttpClient httpClient,
        ObjectMapper objectMapper,
        String baseUrl,
        String apiKey
    ) {
        super(httpClient, objectMapper, baseUrl, apiKey, "/extract");
    }

    /**
     * Extract text from document.
     *
     * @param file Document file
     * @param extractTables Whether to extract tables
     * @param extractImages Whether to extract images
     * @return Extraction results
     * @throws ApiException if the API returns an error
     * @throws IOException if there's a network or I/O error
     */
    public Map<String, Object> extractText(File file, boolean extractTables, boolean extractImages)
            throws ApiException, IOException {
        Map<String, String> additionalFields = new HashMap<>();
        additionalFields.put("extract_tables", String.valueOf(extractTables));
        additionalFields.put("extract_images", String.valueOf(extractImages));

        JsonNode response = postMultipart("/text", file, additionalFields);
        return parseData(response, new TypeReference<Map<String, Object>>() {});
    }

    /**
     * Extract text from document with default options.
     *
     * @param file Document file
     * @return Extraction results
     * @throws ApiException if the API returns an error
     * @throws IOException if there's a network or I/O error
     */
    public Map<String, Object> extractText(File file) throws ApiException, IOException {
        return extractText(file, true, false);
    }

    /**
     * Extract tables from document.
     *
     * @param file Document file
     * @return Extracted tables
     * @throws ApiException if the API returns an error
     * @throws IOException if there's a network or I/O error
     */
    public Map<String, Object> extractTables(File file) throws ApiException, IOException {
        JsonNode response = postMultipart("/tables", file, null);
        return parseData(response, new TypeReference<Map<String, Object>>() {});
    }
}